python/cudf/cudf/_lib/interop.pyx

```diff
@@ -99,31 +100,53 @@ cdef void dlmanaged_tensor_pycapsule_deleter(object pycap_obj):
     dlpack_tensor.deleter(dlpack_tensor)
 
 
-def to_arrow(Table input_table, object column_names, bool keep_index=True):
+cdef vector[column_metadata] gather_metadata(object metadata) except *:
+    """
+    Metadata is stored as lists, and expected format is as follows,
+    [["a", [["b"], ["c"], ["d"]]],       [["e"]],        ["f", ["", ""]]].
+    First value signifies name of the main parent column,
+    and adjacent list will signify child column.
+    """
+    cdef vector[column_metadata] cpp_metadata
+    if isinstance(metadata, list):
+        cpp_metadata.reserve(len(metadata))
+        for i, val in enumerate(metadata):
+            cpp_metadata.push_back(column_metadata(str.encode(str(val[0]))))
+            if len(val) == 2:
+                cpp_metadata[i].children_meta = gather_metadata(val[1])
+
+        return cpp_metadata
+    else:
+        raise ValueError("Malformed metadata has been encountered")
+
+
+def to_arrow(Table input_table,
+             object metadata,
+             bool keep_index=True):
     """Convert from cudf Table to PyArrow Table.
 
     Parameters
     ----------
     input_table : cudf table
     column_names : names for the pyarrow arrays
+    field_names : field names for nested type arrays
     keep_index : whether index needs to be part of arrow table
 
     Returns
     -------
     pyarrow table
     """
 
-    cdef vector[string] cpp_column_names
+    cdef vector[column_metadata] cpp_metadata = gather_metadata(metadata)
     cdef table_view input = (
```

Comment:
```suggestion
    cdef table_view input_table_view = (
```

This is totally allowed, but can we name this `input_table_view`. Just to be on the safer side: https://stackoverflow.com/a/20670757